subsys/bluetooth/host/classic/hfp_ag_internal.h

```diff
@@ -0,0 +1,136 @@
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* HFP AG Supported features */
+#define BT_HFP_AG_SUPPORTED_FEATURES (BT_HFP_AG_FEATURE_INBAND_RINGTONE | BT_HFP_AG_FEATURE_EXT_ERR)
+
+/* bt_hfp_ag flags: the flags defined here represent HFP AG parameters */
+enum {
+	BT_HFP_AG_CMEE_ENABLE,   /* Extended Audio Gateway Error Result Code */
+	BT_HFP_AG_CMER_ENABLE,   /* Indicator Events Reporting */
+	BT_HFP_AG_CLIP_ENABLE,   /* Calling Line Identification notification */
+	BT_HFP_AG_INBAND_RING,   /* In-band ring */
+	BT_HFP_AG_COPS_SET,      /* Query Operator selection */
+	BT_HFP_AG_INCOMING_CALL, /* Incoming call */
+	BT_HFP_AG_CODEC_CONN,    /* Codec connection is ongoing */
+	BT_HFP_AG_CODEC_CHANGED, /* Codec Id Changed */
+
+	/* Total number of flags - must be at the end of the enum */
+	BT_HFP_AG_NUM_FLAGS,
+};
+
+/* HFP HF Indicators */
+enum {
+	HFP_HF_ENHANCED_SAFETY_IND = 1, /* Enhanced Safety */
+	HFP_HF_BATTERY_LEVEL_IND = 2,   /* Remaining level of Battery */
+	HFP_HF_IND_MAX
+};
+
+typedef enum __packed {
+	/** Session disconnected */
+	BT_HFP_DISCONNECTED,
+	/** Session in connecting state */
+	BT_HFP_CONNECTING,
+	/** Session in config state, HFP configuration */
+	BT_HFP_CONFIG,
+	/** Session ready for upper layer traffic on it */
+	BT_HFP_CONNECTED,
+	/** Session in disconnecting state */
+	BT_HFP_DISCONNECTING,
+} bt_hfp_state_t;
+
+typedef enum __packed {
+	/** Call terminate */
+	BT_HFP_CALL_TERMINATE,
+	/** Call outgoing */
+	BT_HFP_CALL_OUTGOING,
+	/** Call incoming */
+	BT_HFP_CALL_INCOMING,
+	/** Call alerting */
+	BT_HFP_CALL_ALERTING,
+	/** Call active */
+	BT_HFP_CALL_ACTIVE,
+	/** Call hold */
+	BT_HFP_CALL_HOLD,
+} bt_hfp_call_state_t;
+
+#define AT_COPS_OPERATOR_MAX_LEN 16
+
+struct bt_hfp_ag {
+	struct bt_rfcomm_dlc rfcomm_dlc;
+	char buffer[HF_MAX_BUF_LEN];
+	uint32_t hf_features;
+	uint32_t ag_features;
+
+	/* HF Supported Codec Ids*/
+	uint32_t hf_codec_ids;
+	uint8_t selected_codec_id;
+
+	ATOMIC_DEFINE(flags, BT_HFP_AG_NUM_FLAGS);
+
+	/* ACL Connection Object */
+	struct bt_conn *acl_conn;
+
+	/* HFP Connection state */
+	bt_hfp_state_t state;
+
+	/* HFP Call state */
+	bt_hfp_call_state_t call_state;
+
+	/* Delayed work deferred tasks:
+	 * - call status cleanup.
+	 */
+	struct k_work_delayable deferred_work;
+
+	/* AG Indicators */
+	uint8_t indicator_value[BT_HFP_AG_IND_MAX];
+	uint32_t indicator;
+
+	/* HF Indicators */
+	uint32_t hf_indicators_of_ag;
+	uint32_t hf_indicators_of_hf;
+	uint8_t hf_indicator_value[HFP_HF_IND_MAX];
+
+	/* operator */
+	char operator[AT_COPS_OPERATOR_MAX_LEN + 1];
+
+	/* SCO Connection Object */
+	struct bt_sco_chan sco_chan;
+	/* HFP TX pending */
+	sys_slist_t tx_pending;
+
+	/* Dial number or incoming number */
+	char number[CONFIG_BT_HFP_AG_PHONE_NUMBER_MAX_LEN + 1];
+
+	/* Calling Line Identification notification
+	 */
+	struct k_work_delayable ringing_work;
+
+	/* Critical locker */
+	struct k_mutex mutex;
```

Comment:
I'd consider using a k_sem instead (initialize it to 1, max 1). You could create helpers like `hfp_ag_lock(struct bt_hfp_ag *ag)` to make the usage more lock-like. Preferring k_sem over k_mutex is at least a general recommendation that's been given by the kernel maintainer.

Reply:
In fact, I don't understand why k_mutex is not used for resource protection. k_sem is mainly used to synchronize the number of resources between threads. If we set max to 1, we can also approximately use it instead of k_mutex. But there is a reason why k_sem cannot replace k_mutex is that k_mutex supports nested resource protection.

Reply:
> In fact, I don't understand why k_mutex is not used for resource protection. k_sem is mainly used to synchronize the number of resources between threads. If we set max to 1, we can also approximately use it instead of k_mutex. But there is a reason why k_sem cannot replace k_mutex is that k_mutex supports nested resource protection, although this situation has not been encountered in the current implementation.

If we get into re-entrancy then I'd argue that we need to look into simplifying the design. Anyway, as you say we're not there right now, and k_sem is a more lightweight solution than k_mutex. It's just a pity that Zephyr doesn't have a lock-style wrapper for it, so you need to do it yourself. @andyross provided a pretty good overview of the options last week in his ZDS talk: https://youtu.be/BDDl9YdZUbQ

Reply:
updated